### PR TITLE
const checking: properly compute the set of transient locals

### PR DESCRIPTION
**Pull request by RalfJung**
For const-checking the MIR of a const/static initializer, we have to know the set of "transient" locals. The reason for this is that creating a mutable (or interior mutable) reference to a transient local is "safe" in the sense that this reference cannot possibly end up in the final value of the const -- even if it is turned into a raw pointer and stored in a union, we will see that pointer during interning and reliably reject it as dangling.

So far, we determined the set of transient locals as "locals that have a `StorageDead` somewhere". But that's not quite right -- if we had MIR like
```rust
StorageLive(_5);
StorageDead(_5);
StorageLive(_5);
// no further storage annotations for _5
```
Then we'd consider `_5` to be "transient", but it is not actually transient.

We do not currently generate such MIR, but I feel uneasy relying on subtle invariants like this. So this PR implements a proper analysis for computing the set of "transient" locals: a local is "transient" if it is guaranteed dead at all `Return` terminators.

Cc @cjgillot

### CODE REVIEW

**Comment by RalfJung** [comment]
Cc @rust-lang/wg-const-eval

**Comment by RalfJung** [comment]
Now I just hope this is okay from a perf perspective... we don't seem to have a cache for these analyses so unfortunately it seems like we have to re-do them here.

@bors try
@rust-timer queue

**Comment by RalfJung** [comment]
Awesome, perf is happy :)

File: compiler/rustc_const_eval/src/check_consts/check.rs [hunk 1]
```diff
@@ -264,23 +268,47 @@ impl<'mir, 'tcx> Checker<'mir, 'tcx> {
...
                 }
-                let mut v = StorageDeads { locals: BitSet::new_empty(ccx.body.local_decls.len()) };
-                v.visit_body(ccx.body);
-                v.locals
+
+                let mut v = TransientLocalVisitor {
+                    maybe_storage_live,
+                    transient: BitSet::new_filled(ccx.body.local_decls.len()),
+                };
+                v.visit_body(&ccx.body);
```
Comment (cjgillot) [inline compiler/rustc_const_eval/src/check_consts/check.rs:309]: A loop over basic blocks may be simpler than a visitor here. 

You may need to filter out unreachable blocks, as the data flow framework ices when seeking inside those.

Comment (RalfJung) [reply compiler/rustc_const_eval/src/check_consts/check.rs:309]: Ah, is that why [this code](https://github.com/rust-lang/rust/blob/84ac80f1921afc243d71fd0caaa4f2838c294102/compiler/rustc_mir_transform/src/lint.rs#L39-L41) only iterates over `traversal::reachable` blocks?

Is there any way to ask the data flow cursor which blocks are live? It seems quite wasteful to re-compute that information using `traversal::reachable`.

Comment (Jaic1) [reply compiler/rustc_const_eval/src/check_consts/check.rs:309]: > Is there any way to ask the data flow cursor which blocks are live?

Seems like improbable, the `ResultCursor` has a field `reachable_blocks` [here](https://github.com/Jaic1/rust/blob/c8b14ba7b6777dfbfb64a02d7bbe867a54dddf2a/compiler/rustc_mir_dataflow/src/framework/cursor.rs#L34) but only enabled with `#[cfg(debug_assertions)]`.

> It seems quite wasteful to re-compute that information using `traversal::reachable`.

It seems just fine to use `traversal::reachable` to do a forward computation once again, considering the dataflow engine might already iterated the blocks multiple times.

Comment (RalfJung) [reply compiler/rustc_const_eval/src/check_consts/check.rs]: All right, I made this do a loop over `traversal::reachable` rather than a visitor.

**Pull request merged**